<a href="https://colab.research.google.com/github/AndrewSLowe/DS-Unit-1-Sprint-3-Statistical-Tests-and-Experiments/blob/master/AL_131_Assignment_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 3 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)
5. You also practice some 1-sample t-tests
6. You try and create some kind of a visualization that communicates the results of your hypothesis tests. This can be as simple as a histogram of the p-values or the t-statistics.

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel
import seaborn as sns
import matplotlib.pyplot as plt

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-09-16 19:36:10--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2019-09-16 19:36:12 (125 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [4]:
# Load Data
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [8]:
df = df.replace({'?':np.NaN, 'n':0, 'y':1})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
# Filter the dataframe into a democrats and republicans dataframes

dem = df[df['party'] == 'democrat']
rep = df[df['party'] == 'republican']

In [14]:
dem.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                         ...                         
democrat                156.0          120.0  ...      160.0         173.0

[1 rows x 16 columns]

In [15]:
rep.groupby(['party']).sum()

handicapped-infants  water-project  ...  duty-free  south-africa
party                                           ...                         
republican                 31.0           75.0  ...       14.0          96.0

[1 rows x 16 columns]

2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01


In [18]:
ttest_ind(dem['handicapped-infants'], rep['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01


In [19]:
ttest_ind(rep['education'], dem['education'], nan_policy='omit')     

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

In [0]:


fix, ax = plt.subplots()
for sample in [stove1, stove2]:
  sns.distplot(sample)

4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)


In [21]:
ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')     

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

5. You also practice some 1-sample t-tests


In [0]:
from scipy.stats import ttest_1samp

In [23]:
len(rep['physician-fee-freeze']) - rep['physician-fee-freeze'].isnull().sum()

165

1) Null Hypothesis:

In 1-sample t-tests YOU GET TO CHOOSE YOUR NULL HYPOTHESIS

𝐻0 : 0.0 - There is ZERO republican support for this bill

2) Alternative Hypothesis

𝐻𝑎 :  𝑥¯≠0  - There is non-zero support for the budget bill among republicans.

3) Confidence Level: 95% or .95

In [24]:
ttest_1samp(rep['physician-fee-freeze'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=115.61141812122193, pvalue=4.438624448197153e-159)

4) t-statistic: 115.61141812122193

5) p-value of .000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000444

--- 

Conclusion: Due to a p-value of 0 I reject the null hypothesis that republican support is zero and conclude that republican support is non-zero. 

In [25]:
ttest_1samp(dem['physician-fee-freeze'], 0, nan_policy='omit')

Ttest_1sampResult(statistic=3.8396428405331067, pvalue=0.0001550891349511097)

4) t-statistic: 3.8396428405331067

5) p-value of .0001550891349511097

--- 

Conclusion: Due to a p-value of 0 I reject the null hypothesis that Democrats support is zero and conclude that republican support is non-zero. 

STRETCH GOAL

Refactor your code into functions so it's easy to rerun with arbitrary variables

In [40]:
dem.head()


party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [0]:
ttest_ind(dem['water-project'], rep['water-project'], nan_policy='omit')     

def t_test(t, r):
  print(ttest_ind(t, r, nan_policy='omit'))